### PR TITLE
PRIME 1762 Site Registration Duplicate PEC

### PR DESCRIPTION
**Pull request by johnnyl6**

### CODE REVIEW

File: prime-angular-frontend/src/app/lib/validators/unique-pec.validator.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import { Injectable } from '@angular/core';
+import { AbstractControl, AsyncValidator, ValidationErrors } from '@angular/forms';
+
+import { Observable } from 'rxjs';
+import { map } from 'rxjs/operators';
+
+import { SiteResource } from '@core/resources/site-resource.service';
+
+@Injectable({ providedIn: 'root' })
```
[the remark this answers is not in the document]
Comment (johnnyl6) [reply prime-angular-frontend/src/app/lib/validators/unique-pec.validator.ts:9]: Neat implementation. Thanks!🙏

**Review by AMiKaH — APPROVED** [review]

File: prime-dotnet-webapi/Services/interfaces/ISiteService.cs [hunk 2]
```diff
@@ -51,5 +51,6 @@ public interface ISiteService
         Task<SiteRegistrationNoteViewModel> GetSiteRegistrationNoteAsync(int siteId, int siteRegistrationNoteId);
         Task<IEnumerable<int>> GetNotifiedSiteIdsForAdminAsync(ClaimsPrincipal user);
         Task<bool> SiteExists(int siteId);
+        Task<bool> PecExistsInNonHealthAuthoritySites(string pec);
```
Comment (james-hollinger) [inline prime-dotnet-webapi/Services/interfaces/ISiteService.cs:54]: I would advise against putting business logic in the name of a method. I makes the name really long and it means that if we want to change the business logic later we have to alter the name. I would prefer something like `PecIsAvailibleAsync` or `ValidatePecAsync` or `CheckPecConflictAsync` or something like that.

File: prime-dotnet-webapi/Services/SiteService.cs [hunk 3]
```diff
@@ -282,6 +287,11 @@ public async Task<Site> UpdatePecCode(int siteId, string pecCode)
             var site = await GetBaseSiteQuery()
                 .SingleOrDefaultAsync(s => s.Id == siteId);
 
+            if (await IsNonHaSiteAndPecNotUnique(site, pecCode))
+            {
+                throw new InvalidOperationException($"Could not update the site.");
```
Comment (james-hollinger) [inline prime-dotnet-webapi/Services/SiteService.cs:292]: We should not be throwing an error in a service method like this due to business logic; especially if the controller is not validating this information. If you try to update the PEC by calling PUT /sites/5/pec and the PEC happens to exist the server will respond with a `500 Internal Service Error` due to this uncaught exception. The controller should be validating this at a minimum.

**Review by mtpultz — APPROVED** [review]
image

**Pull request merged**